R/fit_model.R

```diff
@@ -0,0 +1,179 @@
+    # and below as the R PRNG seed for EpiNow2 initialization
+    seed = seed,
+    warmup = sampler_opts[["iter_warmup"]],
+    samples = sampler_opts[["iter_samples"]],
+    control = list(
+      adapt_delta = sampler_opts[["adapt_delta"]],
+      max_treedepth = sampler_opts[["max_treedepth"]]
+    )
+  )
+
+  df <- data.frame(
+    confirm = data[["confirm"]],
+    date = as.Date(data[["reference_date"]])
+  )
+  rlang::try_fetch(
+    withr::with_seed(seed, {
+      EpiNow2::epinow(
+        df,
+        generation_time = generation_time,
+        delays = delays,
+        truncation = truncation,
+        horizon = horizon,
+        rt = rt,
+        gp = gp,
+        stan = stan,
+        verbose = interactive()
+      )
+    }),
+    # Downgrade model erroring out to a warning so we can catch and return
+    error = function(cnd) {
+      cli::cli_warn(
+        "Model fitting failed. Returning NA.",
+        parent = cnd,
+        class = "failing_fit"
+      )
+      NA
+    }
+  )
+}
+
+#' Format PMFs for EpiNow2
+#'
+#' Opinionated wrappers around EpiNow2::generation_time_opts(),
+#' EpiNow2::delay_opts(), or EpiNow2::dist_spec() that
+#' formats the generation interval, delay, or right truncation parameters
+#' as an object ready for input to EpiNow2.
+#'
+#' Delays or right truncation are optional and
+#' can be skipped by passing `pmf = NA`.
+#'
+#' @param pmf As returned from [CFAEpiNow2Pipeline::read_disease_parameters()].
+#'   A PMF vector or an NA, if not applying the PMF to the model fit.
+#'
+#' @return An EpiNow2::*_opts() formatted object or NA with a message
+#' @name opts_formatter
+NULL
+
+#' @rdname opts_formatter
+#' @export
+format_generation_interval <- function(pmf) {
+  if (
+    rlang::is_na(pmf) || rlang::is_null(pmf)
+  ) {
+    cli::cli_abort("No generation time PMF specified but is required",
+      class = "Missing_GI"
+    )
+  }
+
+  suppressWarnings({
+    EpiNow2::generation_time_opts(
+      dist = EpiNow2::dist_spec(
+        pmf = pmf
+      )
+    )
+  })
+}
+
+#' @rdname opts_formatter
+#' @export
+format_delay_interval <- function(pmf) {
+  if (rlang::is_na(pmf) || rlang::is_null(pmf)) {
+    cli::cli_alert("Not adjusting for infection to case delay")
+    EpiNow2::delay_opts()
+  } else {
+    suppressWarnings({
+      EpiNow2::delay_opts(
+        dist = EpiNow2::dist_spec(
+          pmf = pmf
+        )
+      )
+    })
+  }
+}
+
+#' @inheritParams fit_model
+#' @rdname opts_formatter
+#' @export
+format_right_truncation <- function(pmf, data) {
+  if (
+    rlang::is_na(pmf) || rlang::is_null(pmf)
+  ) {
+    cli::cli_alert("Not adjusting for right truncation")
+    EpiNow2::trunc_opts()
+  } else if (length(pmf) > nrow(data)) {
+    # Nasty bug we ran into where **left-hand** side of the PMF was being
+    # silently removed if length of the PMF was longer than the data,
+    # effectively eliminating the right-truncation correction
```

Comment:
Is this a bug at the `EpiNow2` level? Perhaps add an issue there if there isn't one already

Reply:
(Seems like yes this has been addressed in `EpiNow2`)

Reply:
If so, then for "It doesn't allow a right-truncation PMF longer than the data" if `EpiNow2` is now handling this correctly do we still want to be not allowing this? Naively I don't see a reason why having a right truncation PMF longer than the data is so unreasonable a situation

Reply:
> Is this a bug at the EpiNow2 level

Yes

> (Seems like yes this has been addressed in EpiNow2)

[Yes](https://github.com/epiforecasts/EpiNow2/issues/735)

> " if EpiNow2 is now handling this correctly do we still want to be not allowing this

The latest versions of EpINow2 are handling this correctly, but the version we are running does not. Our version is frozen in a pre-fix state.

Reply:
I guess this makes me question why we are frozen in a pre-fix state (but this is getting off topic)

Reply:
#9

Reply:
Plan is to bump to latest once we have an MVP of this pipeline in place